internal/primitives/api/api.go

```diff
@@ -0,0 +1,84 @@
+// Copyright 2025 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package api
+
+import (
+	"github.com/ChainSafe/gossamer/internal/primitives/core"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	statemachine "github.com/ChainSafe/gossamer/internal/primitives/state-machine"
+	"github.com/ChainSafe/gossamer/internal/primitives/state-machine/overlayedchanges"
+	ptrie "github.com/ChainSafe/gossamer/internal/primitives/trie"
+	"github.com/ChainSafe/gossamer/internal/primitives/trie/recorder"
+)
+
+type ProvideRuntimeApi[
+	N runtime.Number,
+	E runtime.Extrinsic,
+	H runtime.Hash,
+	Hasher runtime.Hasher[H],
+	Backend statemachine.Backend[H, Hasher],
+	Result any,
+	Api ApiExt[N, E, H, Hasher, Backend, Result],
+] interface {
+	// Returns the runtime api.
+	// The returned instance will keep track of modifications to the storage. Any successful
+	// call to an api function, will `commit` its changes to an internal buffer. Otherwise,
+	// the modifications will be `discarded`. The modifications will not be applied to the
+	// storage, even on a `commit`.
+	RuntimeApi() Api
+}
```

Comment:
Here's what I have on my branch, I don't think you need all these generic parameters for this interface.  
```suggestion
// ProvideRuntimeAPI is the interface that provides a runtime api.
type ProvideRuntimeAPI[API any] interface {
	// Returns the runtime api.
	// The returned instance will keep track of modifications to the storage. Any successful call to an api function,
	// will commit its changes to an internal buffer. Otherwise, the modifications will be discarded. The modifications
	// will not be applied to the storage, even on a commit.
	RuntimeAPI() API
}
```

Reply:
That looks simpler, but the problem we're going to have with that is that we need to know which methods we can call on API. That's the reason why I'm using `Api ApiExt[N, E, H, Hasher, Backend, Result]` with the `ApiExt` interface. Did you consider that?

Reply:
When you use `ProvideRuntimeAPI` you can specify `ApiExt` as the generic `API` param.

Reply:
Yeah sure, but what I mean is, don't we need to define the constraint in order to limit the generic to some `ApiExt` implementation?

Reply:
simplified